File: packages/cli/src/ckpt/core/git.py
"""Git subprocess wrapper."""

from __future__ import annotations

import subprocess


def run(
    *args: str,
    capture: bool = False,
    check: bool = True,
) -> subprocess.CompletedProcess[str]:
    """Run a git command.

    Args:
        args: Arguments to pass to git (e.g. "add", ".", "--all").
        capture: If True, capture stdout/stderr instead of streaming to terminal.
        check: If True, raise on non-zero exit code.

    Returns:
        CompletedProcess with returncode and optionally stdout/stderr.
    """
    cmd = ["git", *args]
    kwargs = {"text": True, "check": check}
    if capture:
        kwargs["stdout"] = subprocess.PIPE
        kwargs["stderr"] = subprocess.PIPE
    return subprocess.run(cmd, **kwargs)


def get_repo_root() -> str:
    """Return the absolute path to the repository root."""
    result = run("rev-parse", "--show-toplevel", capture=True)
    return result.stdout.strip()


def get_staged_files() -> list[str]:
    """Return list of staged file paths."""
    result = run("diff", "--cached", "--name-only", capture=True)
    return [f for f in result.stdout.strip().splitlines() if f]


def get_head_hash() -> str:
    """Return the current HEAD commit hash."""
    result = run("rev-parse", "HEAD", capture=True)
    return result.stdout.strip()


def get_parent_hash() -> str:
    """Return the parent commit hash, or empty string for initial commit."""
    try:
        result = run("rev-parse", "HEAD~1", capture=True, check=False)
        if result.returncode == 0:
            return result.stdout.strip()
    except Exception:
        pass
    return ""


def get_author() -> str:
    """Return 'Name <email>' of the current git user."""
    name = run("config", "user.name", capture=True, check=False).stdout.strip()
    email = run("config", "user.email", capture=True, check=False).stdout.strip()
    if name and email:
        return f"{name} <{email}>"
    return name or email or "unknown"


def get_committed_files(commit_hash: str) -> list[str]:
    """Return list of files changed in a specific commit."""
    result = run("diff-tree", "--no-commit-id", "--name-only", "-r", commit_hash, capture=True)
    return [f for f in result.stdout.strip().splitlines() if f]


def passthrough(args: tuple[str, ...]) -> int:
    """Run a git command, streaming output directly to the terminal.

    Returns the exit code.
    """
    result = subprocess.run(["git", *args])
    return result.returncode

File: packages/cli/src/core/git.ts
import { execaSync } from 'execa';

export function run(
  args: string[],
  options?: { capture?: boolean }
): { stdout: string; stderr: string; exitCode: number } {
  try {
    const result = execaSync('git', args, {
      stdio: options?.capture ? 'pipe' : 'inherit',
    });
    return {
      stdout: result.stdout ?? '',
      stderr: result.stderr ?? '',
      exitCode: result.exitCode ?? 0,
    };
  } catch (error: any) {
    if (error.exitCode !== undefined) {
      return {
        stdout: error.stdout ?? '',
        stderr: error.stderr ?? '',
        exitCode: error.exitCode,
      };
    }
    throw error;
  }
}

export function passthrough(args: string[]): number {
  const result = run(args);
  return result.exitCode;
}

export function getRepoRoot(): string {
  const { stdout } = run(['rev-parse', '--show-toplevel'], { capture: true });
  return stdout.trim();
}

export function getStagedFiles(): string[] {
  const { stdout } = run(['diff', '--cached', '--name-only'], { capture: true });
  return stdout.trim() ? stdout.trim().split('\n') : [];
}

export function getHeadHash(): string {
  const { stdout } = run(['rev-parse', 'HEAD'], { capture: true });
  return stdout.trim();
}

export function getParentHash(): string {
  const { stdout, exitCode } = run(['rev-parse', 'HEAD~1'], { capture: true });
  return exitCode === 0 ? stdout.trim() : '';
}

export function getAuthor(): string {
  const { stdout: name } = run(['config', 'user.name'], { capture: true });
  const { stdout: email } = run(['config', 'user.email'], { capture: true });
  const n = name.trim();
  const e = email.trim();
  if (n && e) return `${n} <${e}>`;
  return n || e || 'unknown';
}

export function getCommittedFiles(commitHash: string): string[] {
  const { stdout } = run(
    ['diff-tree', '--no-commit-id', '--name-only', '-r', commitHash],
    { capture: true }
  );
  return stdout.trim() ? stdout.trim().split('\n') : [];
}

export function getRepoUrl(): string {
  try {
    const { stdout, exitCode } = run(['remote', 'get-url', 'origin'], { capture: true });
    return exitCode === 0 ? stdout.trim() : '';
  } catch {
    return '';
  }
}
